### PR TITLE
feat: Google Workspace Toolkit

### PR DESCRIPTION
**Pull request by jtorreggiani**
## Summary of Changes

Implements a toolkit for interfacing with Google Workspace tools like email and calendar.

- Handles the OAuth flow on first tool use
- Simply lists recent emails right now
- List today's Google calendar events

### Requirements

You must have a valid credentials file `~/.config/goose/google_credentials.json` which is generated in the Google developer console.

Demo video:

Listing emails

https://github.com/user-attachments/assets/5effe8f2-5cdc-473c-a633-ceb773270719

Getting today's events
Screenshot 2024-11-06 at 7 18 43 PM

### CODE REVIEW

File: src/goose/tools/google_calendar_client.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+from datetime import datetime, timedelta
+from typing import Any, Dict, List
+from zoneinfo import ZoneInfo
+
+from googleapiclient.discovery import build
+from googleapiclient.errors import HttpError
+
+
+class GoogleCalendarClient:
```
Comment (jtorreggiani) [inline src/goose/tools/google_calendar_client.py:9]: Working on corresponding tests for the calendar client.

**Comment by michaelneale** [comment]
thanks @jtorreggiani - if you have a chance, could you run something like this with goose: 

G> "can you look at the     "google-auth-oauthlib>=1.0.0",
    "google-auth-httplib2>=0.1.0",
    "google-api-python-client>=2.86.0" dependencies and see what size they add to the project" - it will try a few ways, am curious

but otherwise looks good!

**Review by michaelneale — APPROVED** [review]
I think this is great - only concern would be overhead of adding google libs but I expect they are small (and license checks out)

Super cool!

**Comment by Kvadratni** [comment]
> I think this is great - only concern would be overhead of adding google libs but I expect they are small (and license checks out)
> 
> Super cool!

@michaelneale  I ran some and goose insists that it ain't great

```
Let me analyze the results:                                     

 1 Base Python environment size: 64K                            
 2 Environment with Google dependencies: 87M                    
 3 Actual incremental size: ~87MB                               

The dependencies add:                                           

 • 3 direct dependencies                                        
 • 19 transitive dependencies                                   
 • Total of 22 new packages                                     

Key packages contributing to size:                              

 • google-api-python-client and its dependencies                
 • Authentication libraries (google-auth, google-auth-oauthlib) 
 • HTTP and OAuth related libraries                             
 • Protocol buffer libraries                                    

The true incremental impact is approximately 87MB since the base
environment was essentially empty. This represents the full size
impact of adding these Google API dependencies.   
```

**Comment by jtorreggiani** [comment]
Thanks @Kvadratni for doing the investigation on the dependency size here!

@michaelneale, does this make this more of a case for a plugin? I'm curious for the plugins repo, do we still have an issue with other dependencies getting pulled in for other toolkits, even if you don't use them?

If so does that mean we need to distribute toolkits as packages or a Github repository and the plugins repo services as as a registry instead for where source code lives?

**Comment by jtorreggiani** [comment]
@Kvadratni, this is what goose had to say when I ran this.

```
The total size of the virtual environment containing only the
specified Google dependencies and their required packages is
approximately 88M.

To break it down:

 1 google-auth-oauthlib (1.0.0)
 2 google-auth-httplib2 (0.1.0)
 3 google-api-python-client (2.86.0)

These packages, along with their dependencies, add about 88M to the
project. This includes not just the Google packages themselves, but
also their required dependencies like requests, urllib3, protobuf,
etc.

The reason we didn't see a noticeable change in the overall project
size (which remained at 390M) is likely because:

 1 The project already includes many of these dependencies, so
   adding the Google packages didn't introduce a lot of new files.
 2 The 88M addition is relatively small compared to the overall
   project size of 390M, making it hard to notice in the total size.
```

Manually running the diff after recreating the env and running uv sync
```
main> du -sh .
287M

jtorreggiani/google-workspace-toolkit>  du -sh .
372M
```
So still reporting 85M bigger 😢

**Comment by michaelneale** [comment]
yeah I see the same gain - odd it is so large. Yes we don't have a good plugin solution now - you can make it an opt dependency with UV, but then someone needs to know to install `goose-ai[google]` - that is one option so it isn't in by default, longer term we will solve ways to dynamically load toolkits anyway. Might be ok for now for this to have it as it is pretty compelling

**Comment by jtorreggiani** [comment]
> yeah I see the same gain - odd it is so large. Yes we don't have a good plugin solution now - you can make it an opt dependency with UV, but then someone needs to know to install `goose-ai[google]` - that is one option so it isn't in by default, longer term we will solve ways to dynamically load toolkits anyway. Might be ok for now for this to have it as it is pretty compelling

@michaelneale making the dependencies optional for now seems like a reasonable compromise for now to get this merged in.

I've updated the dependency and confirmed I can install it locally using `uv add google_workspace`.

Still familiarizing myself with the `pipx` build process, so not sure if this will work in the same way using `pipx install goose-ai[google_workspace]`.

**Review by michaelneale — APPROVED** [review]
nice one @jtorreggiani - I think this is the right way to do optional deps, and yes you will install it I believe via pipx install goose-ai[....] style - which is fine (but we may have to test things out a bit). 

I think this is great, and as we find a better way to have dynamically loaded toolkits will be great to ensure this is ported over/included (along with goose-plugin things) in a more general and open fashion.

**Pull request merged**